lib/rex/proto/dns/resolver.rb

```diff
@@ -0,0 +1,312 @@
+        packet = Net::DNS::Packet.parse(argument.encode)
+      else
+        packet = make_query_packet(argument,type,cls)
+      end
+
+      # Store packet_data for performance improvements,
+      # so methods don't keep on calling Packet#data
+      packet_data = packet.data
+      packet_size = packet_data.size
+
+      # Choose whether use TCP, UDP
+      if packet_size > @config[:packet_size] # Must use TCP
+        @logger.info "Sending #{packet_size} bytes using TCP due to size"
+        method = :send_tcp
+      else # Packet size is inside the boundaries
+        if use_tcp? or !(proxies.nil? or proxies.empty?) # User requested TCP
+          @logger.info "Sending #{packet_size} bytes using TCP due to tcp flag"
+          method = :send_tcp
+        else # Finally use UDP
+          @logger.info "Sending #{packet_size} bytes using UDP"
+          method = :send_udp unless method == :send_tcp 
+        end
+      end
+
+      if type == Net::DNS::AXFR
+        @logger.warn "AXFR query, switching to TCP" unless method == :send_tcp
+        method = :send_tcp
+      end
+
+      ans = self.__send__(method,packet_data)
+
+      unless (ans and ans[0].length > 0)
+        @logger.fatal "No response from nameservers list: aborting"
+        raise NoResponseError
+        return nil
+      end
+
+      @logger.info "Received #{ans[0].size} bytes from #{ans[1][2]+":"+ans[1][1].to_s}"
+      response = Net::DNS::Packet.parse(ans[0],ans[1])
+
+      if response.header.truncated? and not ignore_truncated?
+        @logger.warn "Packet truncated, retrying using TCP"
+        self.use_tcp = true
+        begin
+          return send(argument,type,cls)
+        ensure
+          self.use_tcp = false
+        end
+      end
+
+      return response
+    end
+
+    #
+    # Send request over TCP
+    #
+    # @param packet_data [String] Data segment of DNS request packet
+    # @param prox [String] Proxy configuration for TCP socket
+    #
+    # @return ans [String] Raw DNS reply
+    def send_tcp(packet_data,prox = @config[:proxies])
+      ans = nil
+      length = [packet_data.size].pack("n")
+      @config[:nameservers].each do |ns|
+        begin
+          socket = nil
+          @config[:tcp_timeout].timeout do
+            catch(:next_ns) do
+              begin
+                config = {
+                  'PeerHost' => ns.to_s,
+                  'PeerPort' => @config[:port].to_i,
+                  'Proxies' => prox,
+                  'Context' => @config[:context],
+                  'Comm' => @config[:comm]
+                }
+                if @config[:source_port] > 0
+                  config['LocalPort'] = @config[:source_port]
+                end
+                if @config[:source_host].to_s != '0.0.0.0'
+                  config['LocalHost'] = @config[:source_host] unless @config[:source_host].nil?
+                end
+                socket = Rex::Socket::Tcp.create(config)
+              rescue
```

Comment:
This is actually not rescue Exception, but standard error, which is fine/safe in this case.